pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/AggregationFunctionUtils.java

```diff
@@ -181,37 +162,63 @@ public static int parsePercentile(String percentileString) {
    * @param arguments Arguments to concatenate
    * @return Concatenated String of arguments
    */
-  public static String concatArgs(List<String> arguments) {
-    return (arguments.size() > 1) ? String.join(CompilerConstants.AGGREGATION_FUNCTION_ARG_SEPARATOR, arguments)
-        : arguments.get(0);
+  public static String concatArgs(String[] arguments) {
+    return arguments.length > 1 ? String.join(CompilerConstants.AGGREGATION_FUNCTION_ARG_SEPARATOR, arguments)
+        : arguments[0];
   }
 
   /**
-   * Compiles and returns all transform expressions required for computing the aggregation, group-by
-   * and order-by
-   *
-   * @param brokerRequest Broker Request
-   * @param functionContexts Aggregation Function contexts
-   * @return Set of compiled expressions in the aggregation, group-by and order-by clauses
+   * Collects all transform expressions required for aggregation/group-by queries.
+   * <p>NOTE: We don't need to consider order-by columns here as the ordering is only allowed for aggregation functions
+   *          or group-by expressions.
    */
-  public static Set<TransformExpressionTree> collectExpressionsToTransform(BrokerRequest brokerRequest,
-      AggregationFunctionContext[] functionContexts) {
-
-    Set<TransformExpressionTree> expressionTrees = new LinkedHashSet<>();
-    for (AggregationFunctionContext functionContext : functionContexts) {
-      AggregationFunction function = functionContext.getAggregationFunction();
-      expressionTrees.addAll(function.getInputExpressions());
+  public static Set<TransformExpressionTree> collectExpressionsToTransform(AggregationFunction[] aggregationFunctions,
+      @Nullable TransformExpressionTree[] groupByExpressions) {
+    Set<TransformExpressionTree> expressions = new HashSet<>();
+    for (AggregationFunction aggregationFunction : aggregationFunctions) {
+      //noinspection unchecked
+      expressions.addAll(aggregationFunction.getInputExpressions());
+    }
+    if (groupByExpressions != null) {
+      expressions.addAll(Arrays.asList(groupByExpressions));
     }
+    return expressions;
+  }
 
-    // Extract group-by expressions
-    if (brokerRequest.isSetGroupBy()) {
-      for (String expression : brokerRequest.getGroupBy().getExpressions()) {
-        expressionTrees.add(TransformExpressionTree.compileToExpressionTree(expression));
-      }
+  /**
+   * Creates a map from expression required by the {@link AggregationFunction} to {@link BlockValSet} fetched from the
+   * {@link TransformBlock}.
+   */
+  public static Map<TransformExpressionTree, BlockValSet> getBlockValSetMap(AggregationFunction aggregationFunction,
+      TransformBlock transformBlock) {
+    //noinspection unchecked
+    List<TransformExpressionTree> expressions = aggregationFunction.getInputExpressions();
+    int numExpressions = expressions.size();
+    if (numExpressions == 0) {
+      return Collections.emptyMap();
```

Comment:
Why do we need these special checks for 0 and 1?

IIUC, instead of executing the for loop once for `numExpressions == 1`, you are using a branch. Can we just have the loop? It will be much cleaner unless I am missing the performance benefit of doing it this way. Loop will take care of returning the empty map, map with 1 KV pair or more

Reply:
We can use loop, but that will always create a HashMap, where all the operations are much more expensive than EmptyMap and SingletonMap. Because most of the functions are zero (`COUNT(*)`) or single input expression, this should give better performance.

Reply:
Got it. Thanks